gwcli/utils.py

```diff
@@ -440,7 +440,7 @@ def __getattr__(self, name):
                        "{}".format(self.args[0]))
                 self.data = Response()
                 self.data.status_code = 500
-                self.data._content = '{{"message": "{}" }}'.format(msg)
+                self.data._content = '{{"message": "{}" }}'.format(msg).encode('utf-8')
```

Comment:
Nit: why is this necessary?

Reply:
According to [this commit msg](https://github.com/psf/requests/commit/4decc7986e32bb8f3511df3dd0c9b1c1d57453c1) of requests as shown below. [And the type of native string in Python2 is bytes but in Python3 is text](https://portingguide.readthedocs.io/en/latest/strings.html#the-native-string), so we need to convert it into bytes explicitly in Python3.
"
JSON *must* be encoded using UTF-8, UTF-16 or UTF-32 (see the [RFC][1]; detect the encoding based on the fact that JSON always starts with 2 ASCII characters.

[1]: http://tools.ietf.org/html/rfc4627#section-3
"
Related issue: https://github.com/psf/requests/issues/4879

Reply:
> But Python 3 defaults to `UTF-8` encoding for strings, right [1]?
> 
> [1] https://docs.python.org/3/howto/unicode.html#the-string-type

Yes, you are right @dillaman . Sorry, my previous explanation (especially the Request commit log). which may cause confusion that this issue is related to encoding. But it is not. It is related to obj type.  It's my bad ;-) , just forget my previous explanation now. Let me explain more details about this here.

First,  we need to know that 'str' object type could be converting into 'bytes'  object type via [str.encode() ](https://docs.python.org/3/library/stdtypes.html#str.encode) method.

Then, let's look at the crash error and related code lines of Request below. When it runs line "nullcount = sample.count(_null)", it throws a TypeError exception "TypeError: must be str, not bytes". Look at the Request code lines below, "_null = '\x00'.encode('ascii')" means that '_null' var is a bytes obj type because it calls encode(). The sample var gets its value from response.content, it is 'str' obj type. Then you got "str.count(bytes)" that's why it throws  TypeError exception "TypeError: must be str, not bytes". 
So the solution here is make sample var as a bytes obj type by calling encode() while assigning value to response.content.


Crash error
"...
File "/usr/lib/python3/dist-packages/gwcli/gateway.py", line
868, in ui_command_create
gateway_hostname = api.response.json()['data']
File "/usr/lib/python3/dist-packages/requests/models.py", line
885, in json
encoding = guess_json_utf(self.content)
File "/usr/lib/python3/dist-packages/requests/utils.py", line
871, in guess_json_utf
nullcount = sample.count(_null)
TypeError: must be str, not bytes
"

Request related code lines
"...
[encoding = guess_json_utf(self.content)](https://github.com/psf/requests/blob/master/requests/models.py#L888)
...
[_null = '\x00'.encode('ascii')  # The null is a bytes obj type
...
sample = data[:4]
...
nullcount = sample.count(_null)](https://github.com/psf/requests/blob/master/requests/utils.py#L856-L876)
"

Reply:
The original error was because the command failed and yet because the status code wasn't checked for success, it attempted to decode the invalid response as json. I think you'll find that this works just fine now that you added the check for success on the `/sysinfo/hostname` call.

At least it works as-is on my RHEL 8 / Python 3 environment w/ just the status check.

Reply:
> The original error was because the command failed and yet because the status code wasn't checked for success, it attempted to decode the invalid response as json. I think you'll find that this works just fine now that you added the check for success on the `/sysinfo/hostname` call.
> 
> At least it works as-is on my RHEL 8 / Python 3 environment w/ just the status check.

Yep, I agree that status code checking is indeed necessary here. And normally it will work well.
But, what will happen if we hit the "status_code != 200" path (One way to hit this path is setting api_host = localhost like git msg saying and create a gateway with a non-localhost ip hostname)? I would say it will get the same crash if we don't put encode() at the end of this response.content. Because in the "status_code != 200" path, it also parse json to get message via method response_message(). And this will happen in other api call too not just the gateway creation. 
So we also need to fix this response.content making it as bytes object type to meet Request's requirement and avoid this potential crash.